[Backend] Correctly apply convert_layout to LoadOp's mask

chenyang78 opened the pull request:
We didn't handle layout conversion correctly when we create an async_copy_global_to_local op. This PR fixed the issue (i.e. https://github.com/openai/triton/issues/3468)

jlebar commented:
Thank you for the fix!

Would you be willing to minimize the testcase?  I'm not sure if the IR itself can be minimized (I know these pipeliner tests can take a lot of IR), but at least I don't think we need all of these CHECKs?

chenyang78 commented:
> Thank you for the fix!
> 
> Would you be willing to minimize the testcase? I'm not sure if the IR itself can be minimized (I know these pipeliner tests can take a lot of IR), but at least I don't think we need all of these CHECKs?

Yes, here is the minimized kernel where the ttgir test came from:

```
import triton
import triton.language as tl

@triton.jit
def foo(in_ptrs, offsets, ts_start, tw_start, out_ptrs):
    seq_start = tl.load(offsets)
    offs_n = tl.arange(0, 32)
    off_o = offs_n[:, None] + offs_n[None, :]
    ts_1_ptrs = ts_start + offs_n
    q = tl.load(in_ptrs + off_o)

    out_ptrs = out_ptrs + off_o
    for _ in range(0, seq_start, 32):
        mask_n = offs_n < 2
        ts_1 = tl.load(ts_1_ptrs, mask=mask_n)
        ts_w = tl.load(tw_start + ts_1)
        q = q + ts_w
        acc = tl.dot(q, q)
        tl.store(out_ptrs, acc)
```

The bug-triggering ttgir was further trimmed with triton-reduce. Yeah, it's kinda hard to minimize pipelining-related ttgir.

chenyang78 commented:
> Thank you for the fix!
> 
> Would you be willing to minimize the testcase? I'm not sure if the IR itself can be minimized (I know these pipeliner tests can take a lot of IR), but at least I don't think we need all of these CHECKs?

Sorry for adding another comment. I hit the comment button too soon. 

Regarding CHECKS, I was trying to capture the connections among relevant IR components. I can see the benefit of keeping a small set of CHECKs. Any suggestion to eliminations? Thanks!

jlebar commented:
Since this testcase causes a verifier error, we don't really need *any* CHECKs.  But if you wanted a few, you could check specifically the load and andi operation.  I don't think you need to CHECK that the operands to the andi are correct, though, because it creates this large chain of CHECKs that is fragile to other changes to the compiler (e.g. reordering instructions).

chenyang78 commented:
> Since this testcase causes a verifier error, we don't really need _any_ CHECKs. But if you wanted a few, you could check specifically the load and andi operation. I don't think you need to CHECK that the operands to the andi are correct, though, because it creates this large chain of CHECKs that is fragile to other changes to the compiler (e.g. reordering instructions).

Make sense. Updated the PR. I removed all the CHECKs but the CHECK-LABEL. Thanks!

jlebar commented:
Thank you!

jlebar commented:
Please `pre-commit install` so that we don't have to re-trigger tests due to clang-format silliness.

chenyang78 commented:
@jlebar Hmm, seems we are hitting a pre-existing failure for AMD backend. I checked it locally on an MI-350 machine and saw a similar accuracy failure, although not exact the same:

```
$ git rev-parse HEAD
ea9777d83cf4a29dc5023ab21071df5446a81e28
$ python3 -c 'import torch; print(torch.version.hip)'
6.0.32830-d62f6a171
$ PYTHONPATH=. LD_LIBRARY_PATH=/opt/rocm-6.0.0/lib  pytest ./test/unit/language/test_core.py -k test_atomic_cas
...
test/unit/language/test_core.py:1471:                                                                                                                            
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
                                                                                                                                                                 
args = (<function assert_allclose.<locals>.compare at 0x7f76fda2d080>, array([63., 63., 63., 63., 63., 63., 63., 63., 63., 63...64., 64., 64., 64., 64., 64., 64.
,                                                                                                                                                                
       64., 64., 64., 64., 64., 64., 64., 64., 64., 64., 64.],                                                                                                   
      dtype=float32))                                                                                                                                            
kwds = {'equal_nan': True, 'err_msg': '', 'header': 'Not equal to tolerance rtol=1e-07, atol=0', 'verbose': True}                                                
                                                                                                                                                                 
    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           AssertionError: 
E           Not equal to tolerance rtol=1e-07, atol=0
E           
E           Mismatched elements: 64 / 128 (50%)
E           Max absolute difference: 1.
E           Max relative difference: 0.015625
E            x: array([63., 63., 63., 63., 63., 63., 63., 63., 63., 63., 63., 63., 63.,
E                  63., 63., 63., 63., 63., 63., 63., 63., 63., 63., 63., 63., 63.,
E                  63., 63., 63., 63., 63., 63., 63., 63., 63., 63., 63., 63., 63.,...
E            y: array([64., 64., 64., 64., 64., 64., 64., 64., 64., 64., 64., 64., 64.,
E                  64., 64., 64., 64., 64., 64., 64., 64., 64., 64., 64., 64., 64.,
E                  64., 64., 64., 64., 64., 64., 64., 64., 64., 64., 64., 64., 64.,...

../../../.conda/envs/my-pt-tmp/lib/python3.11/contextlib.py:81: AssertionError

```

What should we do? Could we disable the flaky tests if they were confirmed to be flaky. Thanks!

jlebar commented:
Yes, please send a PR disabling the test on AMD and file a bug for it.  Then I can assign to AMD folks to fix.

Sorry for the trouble.  :(

chenyang78 commented:
> Yes, please send a PR disabling the test on AMD and file a bug for it. Then I can assign to AMD folks to fix.
> 
> Sorry for the trouble. :(

No worries at all :) Filled an issue (https://github.com/openai/triton/issues/3474) and a PR (https://github.com/openai/triton/pull/3475). Thanks!

chenyang78 commented:
@jlebar Hmm, seems the CI is stalled. Do you know what might go wrong? Thanks!

jlebar commented:
CI was having problems last night.  I just re-approved so maybe it will run this time...

jlebar commented:
> Looks the test triggered an assertion failure fixed by https://github.com/openai/triton/pull/3415, which was reverted by https://github.com/openai/triton/pull/3472. Wondering what we would do here?

https://github.com/openai/triton/pull/3472 makes us pipeline fewer loads, but IMO it's a bug in the pipeliner if changing our decision of which loads to pipeline or not results in crashes.  So I think we should fix that bug (though probably differently than the original proposed PR, which sort of let us make a wrong decision and then undid it).

> BTW, wondering if https://github.com/openai/triton/pull/3473 fixed your internal cases?

Unfortunately not.  It fixed one issue but uncovered another.

chenyang78 commented:
> > Looks the test triggered an assertion failure fixed by #3415, which was reverted by #3472. Wondering what we would do here?
> 
> #3472 makes us pipeline fewer loads, but IMO it's a bug in the pipeliner if changing our decision of which loads to pipeline or not results in crashes. So I think we should fix that bug (though probably differently than the original proposed PR, which sort of let us make a wrong decision and then undid it).

I see. Thanks for the explanation. I had a tentative PR, which I closed, to try to fix the assertion failure:

https://github.com/openai/triton/pull/3455

where I removed invalid loadOp(s) from the pipelining candidates. Would you mind to take a look and see if it would be a reasonable fix? In the PR above, I also had some discussion with Pawel, which you might be interested in as well. Thanks!

> 
> > BTW, wondering if #3473 fixed your internal cases?
> 
> Unfortunately not. It fixed one issue but uncovered another.

Yeah, that's unfortunate. One more thing to note is that the issue that this PR is resolving is also related to #3415.

jlebar commented:
> Would you mind to take a look and see if it would be a reasonable fix?

Right, I agree with Pawel's points, what's in the PR doesn't strike me as the best way to solve the issue.

chenyang78 commented:
> > Would you mind to take a look and see if it would be a reasonable fix?
> 
> Right, I agree with Pawel's points, what's in the PR doesn't strike me as the best way to solve the issue.

Thanks for the feedback. Do you have any suggestion on where it would be the best place to tackle the issue? In the meanwhile, I can take another attempt to come up with a patch. Thanks!

jlebar commented:
> Do you have any suggestion on where it would be the best place to tackle the issue?

WDYM "the best place"?

chenyang78 commented:
> > Do you have any suggestion on where it would be the best place to tackle the issue?
> 
> WDYM "the best place"?

Sorry for causing confusion. I meant to ask a better approach to tackle the issue other than the one from #3455. 

I played with the issue a bit and were wondering if the assertion failure here was actually due to the incompleteness of #3472.  Please see my reasoning below.

The failing test looks like this:

```
    scf.for %arg4 = %c0_i32 to %c2_i32 step %c1_i32  : i32 {
      %15 = tt.load %13
      %16 = tt.addptr %14, %15
      %17 = tt.load %16
      %18 = tt.expand_dims %17 {axis = 0 : i32}
      %19 = tt.broadcast %18
      %20 = arith.addf %9, %19
      %21 = triton_gpu.convert_layout %9
      %22 = triton_gpu.convert_layout %20
      %23 = tt.dot %21, %22, %cst
```

```loadOpsToDistanceAndUse``` produces the following ```loadOpToDistAndUse``` content:

```
# loadOp -> (dist, use)
tt.load %16 -> (0, tt.dot %21, %22, %cst)
tt.load %13 -> (1, tt.load %16)
```

where both loads are legitimate to be pipelined. However, with the rolling-back change in #3472, we skip updating the stage info for the first load above (i.e. ```tt.load %16```) due to the early continue at https://github.com/openai/triton/blob/main/lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp#L403

This leave an invalid stage value in the relevant opInfo item for ```tt.load %16```. Consequently, we hit the assertion failure at https://github.com/openai/triton/blob/main/lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp#L434

A quick workaround would be that we also eliminate ```load %16``` from opInfo before we continue the loop at https://github.com/openai/triton/blob/main/lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp#L404

Otherwise, we would always hit the assertion failure because we are left with an incorrect stage for a valid loadOp candidate. 

Any thought? Thanks!

chenyang78 commented:
@jlebar I updated the PR with a new test. It triggers the same issue that this PR tries to fix and is not affected by the change made in #3472. Thanks!

jlebar reviewed (approved):
Looks great, thank you!
